client/src/components/MultiJurisdictionAudit/RoundManagement/OfflineBatchRoundDataEntry.tsx

```diff
@@ -0,0 +1,229 @@
+import React from 'react'
+import { useParams } from 'react-router-dom'
+import { Formik, FormikProps, Field, FieldArray } from 'formik'
+import { Button, HTMLTable } from '@blueprintjs/core'
+import useContestsJurisdictionAdmin from './useContestsJurisdictionAdmin'
+import { IRound } from '../useRoundsAuditAdmin'
+import useOfflineBatchResults, {
+  IOfflineBatchResults,
+} from './useOfflineBatchResults'
+
+interface IProps {
+  round: IRound
+}
+
+const OfflineBatchRoundDataEntry = ({ round }: IProps) => {
+  const { electionId, jurisdictionId } = useParams<{
+    electionId: string
+    jurisdictionId: string
+  }>()
+  const contests = useContestsJurisdictionAdmin(electionId, jurisdictionId)
+  const [batchResults, updateResults, finalizeResults] = useOfflineBatchResults(
+    electionId,
+    jurisdictionId,
+    round.id
+  )
+
+  if (!contests || !batchResults) return null
+
+  // We only support one contest for now
+  const contest = contests[0]
+
+  const { results, finalizedAt } = batchResults
+
+  return (
+    <Formik
+      initialValues={{ results }}
+      enableReinitialize
+      onSubmit={async values => {
+        // TODO validation to prevent partially empty rows
+        // Omit empty rows
+        const filteredResults = values.results.filter(
+          result =>
+            result.batchName !== '' &&
+            Object.values(result.choiceResults).every(
+              value => (value as string | number) !== ''
+            )
+        )
+        await updateResults(filteredResults)
+        // TODO change isSubmitting?
+      }}
+    >
+      {({
+        handleSubmit,
+        values,
+      }: FormikProps<{ results: IOfflineBatchResults['results'] }>) => (
+        <form>
+          <p>
+            When you have examined all the ballots assigned to you, enter the
+            number of votes recorded for each candidate/choice for each batch of
+            audited ballots.
+          </p>
+          <fieldset disabled={!!finalizedAt}>
+            <HTMLTable>
+              <thead>
+                <tr>
+                  <th />
+                  <th>Batch Name</th>
+                  {contest.choices.map(choice => (
+                    <th key={`th-${choice.id}`}>{choice.name}</th>
+                  ))}
+                </tr>
+              </thead>
+              <FieldArray
+                name="results"
+                render={arrayHelpers => (
+                  <tbody>
+                    {values.results.map((_, r) => (
+                      // eslint-disable-next-line react/no-array-index-key
+                      <tr key={`batchName-${r}`}>
```

Comment:
@jonahkagan Why not use the `batchName` field on the result object here?

Reply:
@jonahkagan I recommend you use something like this:

```
                    {values.results.map((result, r) => (
                      <tr key={`batchName-${result.batchName + r}`}>
```
This means you won't get the error about identical batchnames when you create multiple empty rows before filling them in, and it also retains unique keys when the order changes, even when people use the same batch name for multiple batches (which they shouldn't but it needs to work even if they do).